### PR TITLE
Refactor model and metrics

### PR DESCRIPTION
**Pull request by jasonjewik**
I'll come back to add examples later of how to use the code, and I'm really spotty on documentation, but I'm opening the PR so folks can take a look. Primary changes are such:

- Much easier to load presets (which grabs the model, optimizer, and LR scheduler [if it exists])
- Metrics are now their own subpackage
- Transforms are separated from metrics because IMO the metrics should not have to know they are transforming their inputs

### CODE REVIEW

**Comment by jasonjewik** [comment]
See `Untitled.ipynb` which can be run on MINT1 for some examples.

**Comment by jasonjewik** [comment]
TODO: Vision Transformer is broken. Up to you whether I should fix that in this PR or in a future one.

**Comment by prakhar6sharma** [comment]
> TODO: Vision Transformer is broken. Up to you whether I should fix that in this PR or in a future one.

Let's leave this for a future PR.

**Comment by prakhar6sharma** [comment]
@jasonjewik thanks a lot for this crucial PR. I have done an initial review for the `transforms`, `metrics` and `loaders.py` and added some comments. Most of them include actionable suggestions whereas for a few I have raised the problem without going over the solution(because I couldn't think of a solution).

I am starting my review on the `models/` but in the meantime, please address these comments at a time of your convenience.

**Comment by jasonjewik** [comment]
@prakhar6sharma I don't see your comments - perhaps you did not submit the review?

**Review by prakhar6sharma — CHANGES_REQUESTED** [review]

**Comment by prakhar6sharma** [comment]
> @prakhar6sharma I don't see your comments - perhaps you did not submit the review?

My bad, I thought my comments were visible. They should be visible now. Again, I am would be adding more comments targeting the `models/` folder soon followed by `tests/` and other miscellaneous folders.

Once we are done with the requested changes, I would love to do some kind of stress test locally or on `mint1` before approving the PR. Hope this is fine by you.

File: src/climate_learn/loaders.py
Comment (jasonjewik) [reply src/climate_learn/loaders.py:18]: Haha not dumb at all. It's good to enforce consistency. I'll double-check that everything is alphabetized.

File: src/climate_learn/loaders.py
Comment (jasonjewik) [reply src/climate_learn/loaders.py:34]: `LRScheduler` is an alias for `torch.topim.lr_scheduler`, which is a module. The name of the actual learning rate scheduler base class is `_LRScheduler`. I think what you're suggesting is changing
```python
import torch.optim.lr_scheduler as LRScheduler
```
to
```python
from troch.optim.lr_scheduler import _LRScheduler as LRScheduler
```
This can be done. I'll make the changes shortly.

File: src/climate_learn/loaders.py
Comment (jasonjewik) [reply src/climate_learn/loaders.py:288]: The case where `optimizer is None` is for models with no parameters such as climatology and persistence. For example, you can observe this behavior in the [preset loader test](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/tests/loaders/test_presets.py#L64) and in the [trainer test](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/tests/test_trainer.py).

File: src/climate_learn/loaders.py
Comment (jasonjewik) [reply src/climate_learn/loaders.py:340]: Using `Task` was my initial approach. However, the `in_vars` are dynamic. For example, if the user requests `geopotential` in the task arguments, then this code would think there's just one input variable. In reality, the geopotentials for every pressure level are loaded. I believed that changing the task argument code would be beyond the scope of this PR, since it's a problem from the data side.

File: src/climate_learn/metrics/__init__.py
Comment (jasonjewik) [reply src/climate_learn/metrics/__init__.py:2]: Yes, this is to avoid cluttering the namespace. My intention is for users to load the class and functional definitions separately, like in PyTorch. For example,
```python
import climate_learn.metrics as metrics
import climate_learn.metrics.functional as CLF
```
which mirrors
```python
import torch.nn as nn
import torch.nn.functional as F
```

File: src/climate_learn/metrics/utils.py
Comment (jasonjewik) [reply src/climate_learn/metrics/utils.py:11]: It's true that not all metrics use the metainfo. E.g., vanilla MSE does not use any of these. However, I chose to make this class because (1) I think it's reasonable to assume that these variables should be available for any climate dataset, (2) it allowed me to build a unified initialization function signature for the metrics classes.

File: src/climate_learn/trainer.py
Comment (jasonjewik) [reply src/climate_learn/trainer.py:36]: Good point. I'll change this so that if `accelerator` is not specified, then even if CUDA is available, the trainer will not use GPU acceleration. Only if the user explicitly requests GPU acceleration will the trainer do so.

File: src/climate_learn/loaders.py
Comment (prakhar6sharma) [reply src/climate_learn/loaders.py:34]: > `LRScheduler` is an alias for `torch.topim.lr_scheduler`, which is a module. The name of the actual learning rate scheduler base class is `_LRScheduler`.

I didn't knew that. 




> 
> ```python
> from troch.optim.lr_scheduler import _LRScheduler as LRScheduler
> ```

That would be perfect.

File: src/climate_learn/loaders.py
Comment (jasonjewik) [reply src/climate_learn/loaders.py:88]: How should I inspect whether the datasets have been setup? I don't see a method in `DataModule` that would make this easy.

File: src/climate_learn/loaders.py
Comment (prakhar6sharma) [reply src/climate_learn/loaders.py:340]: Actually when the `.setup()`[[See this](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/src/climate_learn/data/dataset/map_dataset.py#L105)] is called for the `MapDataset` and `ShardDataset`, it first calls the `.setup()` [[See this](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/src/climate_learn/data/dataset/map_dataset.py#L106)] for `climate_dataset` which returns `variables_to_update`.

`variables_to_update` is a dict where the keys are the old variable names (like `geopotential` in this case) and values are a list of new variable names corresponding to the old one (like `['geopotential_50', 'geopotential_250', ..., etc]` in this case). Then, the `.setup()` corresponding to the `MapDataset` and `ShardDataset` calls the `.setup()`[[See this](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/src/climate_learn/data/dataset/map_dataset.py#L122)] corresponding to the `task`. Which updates the `in_vars`, `out_vars` and `constants` for the `Task`.

Thus, if the `.setup() `has been called for `dataset` then you can safely use the `in_vars`. In fact the current logic would also fail if the `.setup()` hasn't been called because there would be no data in the memory, thus `dataloader` would give an error.

File: src/climate_learn/loaders.py
Comment (jasonjewik) [reply src/climate_learn/loaders.py]: This is handled in `LitModule`: [training phase](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/src/climate_learn/models/module.py#L40), [validation/testing phases](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/src/climate_learn/models/module.py#L73).

File: src/climate_learn/metrics/metrics.py
Comment (jasonjewik) [reply src/climate_learn/metrics/metrics.py:48]: Yes, it would throw an error. Writing `*args, **kwargs` was merely for my convenience. I can change this to explicitly say `aggregate_only` and `metainfo`.

File: src/climate_learn/metrics/metrics.py
Comment (jasonjewik) [reply src/climate_learn/metrics/metrics.py:103]: Good catch - I missed that. I'll update the type hint in the initialization signature too.

File: src/climate_learn/loaders.py
Comment (prakhar6sharma) [reply src/climate_learn/loaders.py:88]: Well for now, you can call the `get_lat_lon()` as it would return `None`. Once, this PR is merged, I would open a new PR which would then add a method corresponding to this check. I would also then change the logic from `get_lat_lon()` to that new method.

File: src/climate_learn/metrics/__init__.py
Comment (prakhar6sharma) [reply src/climate_learn/metrics/__init__.py:2]: I see. That makes sense.

File: src/climate_learn/metrics/metrics.py
Comment (jasonjewik) [reply src/climate_learn/metrics/metrics.py:198]: TBH, not sure. I assumed our old implementation (with the for-loops) to be correct, so when I parallelized the code here, I just did whatever made the new implementation results match the old implementation results. If that answer's not satisfying enough, I can dig around for a proof of correctness. :sweat_smile:

File: tests/loaders/utils.py
Comment (prakhar6sharma) [inline tests/loaders/utils.py:56]: This line should be removed.

File: tests/loaders/utils.py
Comment (prakhar6sharma) [inline tests/loaders/utils.py:57]: This line too since the `get_lat_lon()` has a return signature of `Tuple[Union[np.ndarray, None], Union[np.ndarray, None]]`.

File: tests/loaders/utils.py
Comment (prakhar6sharma) [inline tests/loaders/utils.py:52]: `get_climatology()` should return `None` if `.setup()` hasn't been called yet.

File: tests/loaders/utils.py
Comment (prakhar6sharma) [inline tests/loaders/utils.py:49]: `train_datalaoder()` should raise some kind of error and **not return `self.batches`** if `.setup()` hasn't been called yet.

**Review by prakhar6sharma — COMMENTED** [review]
I have finished my review and leaved some comments. Would be great if you could address them.

**Review by prakhar6sharma — CHANGES_REQUESTED** [review]
I have finished my review and added some comments. It would be great if you could address them.

**Comment by prakhar6sharma** [comment]
> TODO: Vision Transformer is broken. Up to you whether I should fix that in this PR or in a future one.

Based on my current understanding of this PR, seems like you have already fixed it. Am I correct here?

File: src/climate_learn/loaders.py
Comment (prakhar6sharma) [inline src/climate_learn/loaders.py:87]: I think based on the import this can be changed from `LRScheduler._LRScheduler` to `LRScheduler`.

File: src/climate_learn/loaders.py
Comment (prakhar6sharma) [reply src/climate_learn/loaders.py:335]: Can you please update the logic for `get_data_dims`, such that we don't make a call to `train_dataloader()`. My reason for asking so, is because of the performance reasons for `ShardDataset` i.e. we would need to load a chunk into memory.

I believe, in the current PR we can assume that the `.setup()` has been called for the `train_dataset()`. (**please also add this assumption as a comment, so that we can later revisit this issue and remove this assumption and add corresponding error handling code**). If we make this assumption, then we can use `Task` and logic inside `collate_fn() ` to return the **input** (`x`) and **output** (`y`) shapes.

**Comment by prakhar6sharma** [comment]
Other than these changes, I feel that each component such as `Climatology`, `Interpolation`, `LinearRegression`, `ResNet`, etc should have an attribute named `in_vars` and `out_vars` other than `in_channels` and `out_channels`.

I believe that it would allow us to implement some crucial checks such as in the case of  `Persistence`, we need the `out_vars` to be subset of `in_vars`. Using just `in_channels` and `out_channels` it's not possible to do so.

File: src/climate_learn/models/hub/interpolation.py
Comment (prakhar6sharma) [inline src/climate_learn/models/hub/interpolation.py:6]: I think the classe should be decorated using `register` (`register("interpolation"`)).

File: src/climate_learn/models/hub/linear_regression.py
Comment (prakhar6sharma) [inline src/climate_learn/models/hub/linear_regression.py:5]: I think the classe should be decorated using register (register("linear_regression")).

File: src/climate_learn/models/hub/persistence.py
Comment (prakhar6sharma) [inline src/climate_learn/models/hub/persistence.py:5]: I think the classe should be decorated using register (register("persistence")).

**Review by prakhar6sharma — CHANGES_REQUESTED** [review]
Spotted some more minor issues.

File: src/climate_learn/metrics/metrics.py
Comment (prakhar6sharma) [reply src/climate_learn/metrics/metrics.py:201]: Reopening this comment. When I thought more about it, I realized that the previous code was indeed incorrect.

By definition of `RMSE`, I found on wikipedia, we first take a `.square()` over the `difference` between `target` and `pred` then take a `.mean()` and finally do a `.sqrt() `.

Thus for the per-channel case, we need to take `mean` over the **entire examples** for every **latitude** and **longitude** for every channel. Hence, it should be `error.mean([0,2, 3]).sqrt()`, in my opinion.

**Review by prakhar6sharma — CHANGES_REQUESTED** [review]
Apologies for adding multiple different reviews.

File: src/climate_learn/loaders.py
Comment (jasonjewik) [reply src/climate_learn/loaders.py:335]: I don't understand what you are asking me to do. `collate_fn` requires a batch anyway, so wouldn't that also require me to use the dataloader?

File: src/climate_learn/models/module.py
Comment (jasonjewik) [reply src/climate_learn/models/module.py:39]: This is necessary because of the models which PyTorch Lightning would not autocast such as climatology. Climatology is set before casting to device happens.

File: src/climate_learn/models/module.py
Comment (jasonjewik) [reply src/climate_learn/models/module.py:46]: We do.

File: src/climate_learn/models/module.py
Comment (jasonjewik) [reply src/climate_learn/models/module.py:69]: Same as my previous comment.

File: src/climate_learn/models/module.py
Comment (jasonjewik) [reply src/climate_learn/models/module.py:74]: You have correctly ascertained the correct behavior. I aim to provide as much flexibility to the user as possible so they can pick any combination of transforms and metrics. Besides, in the future, we should definitely change the data code so that the user can specify whatever transform they want since normalization is not appropriate for all modalities of climate data. I will update my code to make this intended purpose more clear.

**Comment by jasonjewik** [comment]
> Other than these changes, I feel that each `model` under `hub/` such as `Climatology`, `Interpolation`, `LinearRegression`, `ResNet`, etc should have an attribute named `in_vars` and `out_vars` other than `in_channels` and `out_channels`.
> 
> I believe that it would allow us to implement some crucial checks such as in the case of `Persistence`, we need the `out_vars` to be subset of `in_vars`. Using just `in_channels` and `out_channels` it's not possible to do so. Similarly, in the case of `Interpolation` we can ensure that `in_vars` are same as `out_vars`.

I disagree. For clean abstraction of code, the models should not be aware of the variable names. It is the loader functions' responsibility to ensure that the requested model is legal.

File: src/climate_learn/models/module.py
Comment (prakhar6sharma) [reply src/climate_learn/models/module.py:50]: The updated code doesn't log aggregate loss for per-channel setting.

File: tests/loaders/utils.py
Comment (prakhar6sharma) [inline tests/loaders/utils.py:59]: Even though this new test captures the essence of current `DataModule`, but to mimic the exact behavior, the `DataModule` won't raise an error when `get_climatology()` is called, it indeed would return a `None` value. Thus, it becomes the task of the callee to have proper error handling code.

File: src/climate_learn/models/module.py
Comment (prakhar6sharma) [inline src/climate_learn/models/module.py:49]:
```suggestion
        else:  # per channel + aggregate
            loss_dict[f"{loss_name}:aggregate_loss"] = losses[-1]
```

File: src/climate_learn/metrics/functional.py
Comment (prakhar6sharma) [inline src/climate_learn/metrics/functional.py:8]: Is it necessary to import this class because we are not really calling it in this file?

**Review by prakhar6sharma — CHANGES_REQUESTED** [review]
Added some more changes.

File: src/climate_learn/loaders.py
Comment (prakhar6sharma) [reply src/climate_learn/loaders.py:354]: Let's ignore this then for now. I will open a new PR for this once this gets merged.

**Comment by prakhar6sharma** [comment]
> > Other than these changes, I feel that each `model` under `hub/` such as `Climatology`, `Interpolation`, `LinearRegression`, `ResNet`, etc should have an attribute named `in_vars` and `out_vars` other than `in_channels` and `out_channels`.
> > I believe that it would allow us to implement some crucial checks such as in the case of `Persistence`, we need the `out_vars` to be subset of `in_vars`. Using just `in_channels` and `out_channels` it's not possible to do so. Similarly, in the case of `Interpolation` we can ensure that `in_vars` are same as `out_vars`.
> 
> I disagree. For clean abstraction of code, the models should not be aware of the variable names. It is the loader functions' responsibility to ensure that the requested model is legal.

Do we assume that the user would create the models using the loader functions only? 

- If yes, then I completely agree with you. 
- If not, then we are allowing silent bugs to happen. 
-- Let me give an example of `Persistence`, let's say the `len(in_vars)` is same as `len(out_vars)` but `out_vars` is not a subset of `in_vars`. In this case, `Persistence` would work fine but logically the output would be incorrect. 
-- Moreover, we should support `Persistence` for cases where `out_vars` is a subset of `in_vars`. Unfortunately, the current attributes of the class doesn't allow it to know which `channels` of `input` are present in `output` also.
-- Similarly, I can also build an example for `Interpolation`.

**Comment by jasonjewik** [comment]
> Do we assume that the user would create the models using the loader functions only?
> 
> * If yes, then I completely agree with you.
> * If not, then we are allowing silent bugs to happen.
>   -- Let me give an example of `Persistence`, let's say the `len(in_vars)` is same as `len(out_vars)` but `out_vars` is not a subset of `in_vars`. In this case, `Persistence` would work fine but logically the output would be incorrect.
>   -- Moreover, we should support `Persistence` for cases where `out_vars` is a subset of `in_vars`. Unfortunately, the current attributes of the class doesn't allow it to know which `channels` of `input` are present in `output` also.
>   -- Similarly, I can also build an example for `Interpolation`.

Yes. I assume the user to create models with the loader function only. Side-by-side comparison of the number of lines of code required, assuming the data module `dm` has already been created, setup, etc. shows that it's the same.
```python
# using the loaders
import climate_learn as cl
model, optimizer, lr_scheduler = cl.load_preset("forecasting", dm, "persistence")

# do it yourself
from climate_learn.models.hub import Persistence
model = Persistence()
```
Besides, as the user wishes to load more models, the first approach scales better since they don't have to add a new import for each model. If the user chooses to do the second route (load it manually) and something like what you describe happens, IMO it is the user's fault, not the code's fault. Happy to discuss further.

File: tests/loaders/utils.py
Comment (prakhar6sharma) [reply tests/loaders/utils.py:51]: currently the `dataloader` will raise an `error` only. It's the get_climatology which will silently return `None`.

**Review by prakhar6sharma — CHANGES_REQUESTED** [review]

**Comment by prakhar6sharma** [comment]
> Yes. I assume the user to create models with the loader function only. Side-by-side comparison of the number of lines of code required, assuming the data module `dm` has already been created, setup, etc. shows that it's the same.
> 
> ```python
> # using the loaders
> import climate_learn as cl
> model, optimizer, lr_scheduler = cl.load_preset("forecasting", dm, "persistence")
> 
> # do it yourself
> from climate_learn.models.hub import Persistence
> model = Persistence()
> ```
> 
> Besides, as the user wishes to load more models, the first approach scales better since they don't have to add a new import for each model. If the user chooses to do the second route (load it manually) and something like what you describe happens, IMO it is the user's fault, not the code's fault. Happy to discuss further.

Looks good to me. It's just that the `loaders.py` would need to be changed every time you add a model.

Also, can you please add support for `Persistence` for cases where `out_vars` is a subset of `in_vars`.

**Review by prakhar6sharma — APPROVED** [review]

**Review by prakhar6sharma — CHANGES_REQUESTED** [review]
Hopefully, this one should be the last request change.

**Review by prakhar6sharma — CHANGES_REQUESTED** [review]
Hopefully, this one should be the last request change.

**Review by prakhar6sharma — APPROVED** [review]
Looks good to me!!!

**Pull request merged**
